File: CivitasRuby/lib/sorpresa.rb
require_relative 'tipo_sorpresa'

require_relative 'diario'
require_relative 'mazo_sorpresas'
require_relative 'tablero'
require_relative 'jugador'
require_relative 'casilla'

module Civitas
  
  class Sorpresa
    attr_reader :nombre
    
    
    def initialize(texto,valor,mazo,tablero,tipo)
      
      @texto = texto
      @valor = valor
      @mazo = mazo
      @tipo = tipo
      @tablero = tablero
      
      
    end
    
    def self.new_por_tablero(tipo , tablero)
      new("",-1,nil,tablero,tipo)
    end
    
    def self.new_por_tablero_valor(tipo , tablero , valor , texto)
      new(texto,valor,nil,tablero,tipo)
    end
    
    def self.new_por_valor(tipo , valor , texto)
      new(texto,valor,nil,nil,tipo)
    end
    
    def self.new_por_mazo(tipo , mazo)
      new("",-1,mazo,nil,tipo)
    end
    
    def jugador_correcto(actual,todos)
     resultado = false
     if actual < todos.length and actual >= 0
       resultado = true
     end
     resultado  
    end
    
    def aplicar_a_jugador(actual , todos)
      case @tipo
      when TipoSorpresa::IRCARCEL
          aplicar_a_jugador_ir_carcel(actual,todos);
      when TipoSorpresa::IRCASILLA
         aplicar_a_jugador_ir_a_casilla(actual, todos);
      when TipoSorpresa::SALIRCARCEL
          aplicar_a_jugador_salir_carcel(actual, todos);
      when TipoSorpresa::PAGARCOBRAR
          aplicar_a_jugador_pagar_cobrar(actual, todos);
      when TipoSorpresa::PORCASAHOTEL
          aplicar_a_jugador_por_casa_hotel(actual, todos);
      when TipoSorpresa::PORJUGADOR
          aplicar_a_jugador_por_jugador(actual,todos)
        
      end
    end
    
    def salir_del_mazo()
      
      sorpresa = Sorpresa.new_por_tablero(TipoSorpresa::SALIRCARCEL , @tablero)
      
      if(@tipo == TipoSorpresa::SALIRCARCEL)
        @mazo.inhabilitarCartaEspecial(sorpresa)
      end
    end
    
    
    def to_string
      
      "Sorpresa de tipo: "+@tipo.to_s+"\n"
      
    end
    
    def usada
      sorpresa = Sorpresa.new_por_tablero(TipoSorpresa::SALIRCARCEL,@tablero)
      
      if @tipo == TipoSorpresa::SALIRCARCEL
        @mazo.inhabilitarCartaEspecial(sorpresa)
      end
    end
    
    private
    
    def informe(actual , todos)
      Diario.instance.ocurre_evento("Se ha aplicado una sorpresa a "+todos[actual].nombre)
    end
    
    
    def aplicar_a_jugador_ir_carcel(actual,todos)
      
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        todos[actual].encarcelar(@tablero.carcel);
      end
      
    end
    
    def aplicar_a_jugador_ir_a_casilla(actual, todos)
      
      if(jugador_correcto(actual,todos))
        informe(actual,todos)
        aux = todos[actual].num_casilla_actual
        todos[actual].mover_a_casilla(@tablero.nueva_posicion(aux, @tablero.calcularTirada(aux,@valor)))
        @tablero.casilla(@tablero.nueva_posicion(aux,@tablero.calcularTirada(aux, @valor))).recibe_jugador(actual, todos)
      end
    end
      
    end

    def aplicar_a_jugador_salir_carcel(actual, todos)
      jugador = todos[actual]  

      es_correcto = jugador_correcto(actual,todos)
      
      if(es_correcto)
        
        hay_salvoconducto=false
        informe(actual,todos)
        
        for jug in todos
          
          if jug.tiene_salvoconducto
            hay_salvoconducto=true
          end
          
        end
        
        unless(hay_salvoconducto)
          @jugador.obtener_salvoconducto(self)
          salir_del_mazo()
        end
      end
    end
    
    def aplicar_a_jugador_pagar_cobrar(actual, todos)
      jugador = todos[actual]
      es_correcto = jugador_correcto(actual,todos)
      if(es_correcto)
        informe(actual,todos)
        jugador.modificar_saldo(@valor)
      end
    end
    
    def  aplicar_a_jugador_por_casa_hotel(actual, todos)
      
        jugador = todos[actual];
        es_correcto = jugador_correcto(actual, todos);
        
        if es_correcto 
            informe(actual, todos);
            jugador.modificar_saldo(@valor * jugador.cantidad_casas_hoteles());
        end
      
    end
    
    
    def aplicar_a_jugador_por_jugador(actual,todos)
      
      jugador = todos[actual]
      es_correcto = jugador_correcto(actual , todos)
      
      if es_correcto
        
        informe(actual , todos)
        pagar_cobrar1 = Sorpresa.new_por_valor(TipoSorpresa::PAGARCOBRAR , -1*@valor , "cobra")
        pagar_cobrar2 = Sorpresa.new_por_valor(TipoSorpresa::PAGARCOBRAR , (todos.length-1) * @valor , "cobra")
        
        i = 0
        for jug in todos
          
          if jug != jugador
            pagar_cobrar1.aplicar_a_jugador(i,todos)
          end
          
          i = i + 1
          
        end
        
        pagar_cobrar2.aplicar_a_jugador(actual,todos)
        
      end
      
    end
    
end

File: CivitasJava/src/civitas/Sorpresa.java
package civitas;

import java.util.ArrayList;

public class Sorpresa{

    private String texto;
    private int valor;
    
    //asociaciones
    MazoSorpresas mazo;
    TipoSorpresa tipo;
    Tablero tablero;
    
    
    //paquete
    
    Sorpresa(TipoSorpresa tipo , Tablero tablero){//Constructor Enviar Carcel
        this(tipo,tablero,-1,tipo.toString());
    }
    
    Sorpresa(TipoSorpresa tipo , Tablero tablero , int valor , String texto){//Envia otra casilla
        
        
        this.tipo = tipo;
        this.tablero = tablero;
        this.texto = texto;
        this.mazo = null;
        this.valor = valor;
    }
    
    Sorpresa(TipoSorpresa tipo , int valor , String texto){//Resto sorpresas
        
        this(tipo,null,valor,texto);
        
    }
    
    Sorpresa(TipoSorpresa tipo , MazoSorpresas mazo){//EVITAR CARCEL
        init();
        this.tipo = tipo;
        this.texto= tipo.toString();
        this.mazo = mazo; 
        
    }
    
    void aplicarAJugador(int actual , ArrayList<Jugador> todos){
        

        switch(tipo){         
            case IRCARCEL:
                aplicarAJugador_irCarcel(actual,todos);
            case IRCASILLA:
                aplicarAJugador_irACasilla(actual, todos);
            case SALIRCARCEL:
                aplicarAJugador_salirCarcel(actual, todos);
            case PAGARCOBRAR:
                aplicarAJugador_pagarCobrar(actual, todos);
            case PORCASAHOTEL:
                    aplicarAJugador_porCasaHotel(actual, todos);
            case PORJUGADOR:
                aplicarAJugador_porJugador(actual, todos);
        }
           
        
    }
    
    void salirDelMazo(){
       
        if(tipo == TipoSorpresa.SALIRCARCEL)
        {
            mazo.inhabilitarCartaEspecial(this);
        }
        
    }
    
    void usada(){
        
        if(tipo == TipoSorpresa.SALIRCARCEL)
            mazo.habilitarCartaEspecial(this);
        
    }
    
    //public
    public boolean jugadorCorrecto(int actual , ArrayList<Jugador> todos){
        
        return (actual >= 0) && (actual<todos.size());
    }
    
    @Override
    public String toString(){
        
        return "\nnombre de la sorpresa:"+tipo+valor+"\n";
    }
    
    //private
    
    private void init(){
        
        valor = -1;
        mazo = null; 
        tablero = null;
        
    }
    
    private void aplicarAJugador_irACasilla(int actual , ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual, todos);
            
            int casillaActual = todos.get(actual).getNumCasillaActual();

            int tirada=tablero.CalcularTirada(casillaActual, valor);
            
            int nuevaPosicion = tablero.nuevaPosicion(casillaActual, tirada);
            
            todos.get(actual).moverACasilla(nuevaPosicion);
            
            tablero.getCasilla(nuevaPosicion).recibeJugador(actual, todos);
            
        }
        
        
    }
    
    private void aplicarAJugador_irCarcel(int actual , ArrayList<Jugador> todos){
        
        boolean esCorrecto = jugadorCorrecto(actual, todos);
        
        if(esCorrecto){
            informe(actual, todos);
            todos.get(actual).encarcelar(tablero.getCarcel());
        }     
    }
    
    private void aplicarAJugador_pagarCobrar(int actual , ArrayList<Jugador> todos){
        
        Jugador jugador = todos.get(actual);
        boolean esCorrecto = jugadorCorrecto(actual, todos); 
        
        if(esCorrecto){
            informe(actual,todos);
            jugador.modificarSaldo(valor);
        } 
    }
    
    private void aplicarAJugador_porCasaHotel(int actual , ArrayList<Jugador> todos){
        
        Jugador jugador = todos.get(actual);
        boolean esCorrecto = jugadorCorrecto(actual, todos);
        
        if(esCorrecto){
            informe(actual, todos);
            jugador.modificarSaldo(valor * jugador.cantidadCasasHoteles());
            
        }     
    }
        private void aplicarAJugador_porJugador(int actual , ArrayList<Jugador> todos){
        
        Jugador jugador = todos.get(actual);
        boolean esCorrecto = jugadorCorrecto(actual, todos);
        
        if(esCorrecto)
        {
            informe(actual,todos);
            Sorpresa pagarCobrar1 = new Sorpresa(TipoSorpresa.PAGARCOBRAR , -1*valor , "cobra");
            Sorpresa pagarCobrar2 = new Sorpresa(TipoSorpresa.PAGARCOBRAR , (todos.size()-1)*valor , "cobra");
            
            for(int i = 0 ; i < todos.size() ; i++)
            {
                if(todos.get(i) != jugador)
                {
                    pagarCobrar1.aplicarAJugador_pagarCobrar(i, todos);
                }
            }
            
            pagarCobrar2.aplicarAJugador_pagarCobrar(actual, todos);
        }    
        
    }
    
    
    private void aplicarAJugador_salirCarcel(int actual , ArrayList<Jugador> todos){
        
        Jugador jugador = todos.get(actual);
        boolean esCorrecto = jugadorCorrecto(actual, todos);
        
        if(esCorrecto){
            boolean haySalvoConducto = false;
            informe(actual, todos);
            
            for(int i = 0 ; i < todos.size() ; i++){
                if (todos.get(i).tieneSalvoconducto()){
                    haySalvoConducto = true;   
                }
            }
            
            if(!haySalvoConducto){
                jugador.obtenerSalvoConducto(this);
                salirDelMazo();
            }
            
        }
        
    }
    
    private void informe(int actual , ArrayList<Jugador> todos){
        
        Diario.getInstance().ocurreEvento("Se ha aplicado una sorpresa a "
        +todos.get(actual).getNombre());
        
    }
    
    
        public static void main(String [] args){
        
        Jugador j1=new Jugador("Roberto");
        
        Tablero t1=new Tablero(4);
        t1.añadeJuez();
        t1.añadeCasilla(new Casilla(new TituloPropiedad("calle1",200,2,200,500,250)));
        t1.añadeCasilla(new Casilla(new TituloPropiedad("calle2",300,2,200,600,260)));
        
        Sorpresa s1=new Sorpresa(TipoSorpresa.IRCASILLA,t1,3,"ve");
        
        
    }
}
